### DISCUSSION TITLE
Setting X-Robots-Tag for preview with custom domain

### DISCUSSION THREAD

**Discussion by bscaspar** [vercel/vercel#5714] [Help]
The docs note in a few places that the `X-Robots-Tag` header is not set in a preview environment for a custom domain. What would be the recommended way to set this header in this scenario?

**Comment by bscaspar** [comment]
In my `next.config.js` file I added the following config:
```js
webpack: (config, options) => {
    // Generate disallow all robots.txt and add to /public folder in preview
    if (options.isServer && process.env.VERCEL_ENV === 'preview') {
      require('./scripts/generate-robotstxt-preview.js');
    }

//other webpack config here
}
```

This adds the generated `robots.txt` file which disallows all crawlers (the ones that actually follow the `robots` directions) from crawling the preview sites, custom domain or not.

I'm open to suggestions if folks have other methods for this!

**Reply by zachguo** [reply]
Could you share your script? I tried to write a new `robots.txt` into `./public` and `./`, neither worked.
```js
  webpack: (config, options) => {
    if (options.isServer && process.env.VERCEL_ENV === 'preview') {
      const fs = require('fs');
      fs.writeFileSync("./robots.txt", "User-agent: *\nDisallow: /");
    }
    return config;
  }
```

**Reply by bscaspar** [reply]
Sure, it's located in `/scripts` and here's the code. It looks like you may be referencing the `public` directory incorrectly, maybe check if `public/robots.txt` works.
```js
const fs = require('fs');

(() => {
  const robots = `User-agent: * \nDisallow: /`;

  fs.writeFileSync('public/robots.txt', robots);
  console.log('robots.txt written to public/robots.txt');
})();
```

Also after posting my response I did some more research and learned [here](https://developers.google.com/search/docs/advanced/robots/robots-faq) that the best way to guarantee a URL is not indexed is to use the robots meta tag or `X-Robots-Tag` HTTP header instead. So I actually removed the `robots.txt` from preview and added this to the `_app.ts` file: 
```jsx
{(process.env.VERCEL_ENV === 'preview' ||
  process.env.VERCEL_ENV === 'development') && (
  <meta name='robots' content='noindex' />
)}
```

**Reply by rtitlestad** [reply]
The above worked for me although I had to use `NEXT_PUBLIC_VERCEL_ENV `
```javascript
{process.env.NEXT_PUBLIC_VERCEL_ENV === 'preview' && (
  <meta name="robots" content="noindex" />
)}
```

**Reply by leerob** [reply]
Is there a reason you want Preview URLs indexed on Google? I general, I would not recommend doing this. I would only recommend having your site indexed in the Production environment.

**Reply by bscaspar** [reply]
Hey @leerob the goal was to _prevent_ indexing with the `content='noindex'` prop on the meta tag. I have since discovered the [x-robots-tag](https://vercel.com/docs/edge-network/headers?query=robots#x-robots-tag) applied by Vercel in preview and realized I was overthinking this since you all at Vercel did the work for us! 😅

**Reply by rtitlestad** [reply]
@bscaspar you were right though (in your initial question) that the x-robots-tag won't be applied to to a preview deployment on a **custom** domain e.g. when you assign your `develop` branch to `staging.example.com`

**Reply by bscaspar** [reply]
Good point, I guess I turned myself around a bit on this!

**Reply by lukasbals** [reply]
If you use the [`next-sitemap`](https://www.npmjs.com/package/next-sitemap) package you can configure the generation of the `robots.txt` file like this: 

```js
module.exports = {
  siteUrl: 'https://your-domain',
  generateRobotsTxt: true,
  robotsTxtOptions: {
    policies: [
      {
        userAgent: '*',
        [process.env.VERCEL_ENV !== 'preview' && process.env.VERCEL_ENV !== 'development'
          ? 'allow'
          : 'disallow']: '/',
      },
    ],
  },
};
```

**Comment by AdamLantz** [comment]
Very late to this thread, but here's a more direct answer to the question. 
As of Next.js `9.5.0`, you can set [custom response headers](https://nextjs.org/docs/api-reference/next.config.js/headers).

Update your `next.config.js` so that any incoming request will be include `X-Robots-Tag: noindex` in the response headers like this:
```js
module.exports = {
  async headers() {
    const headers = [];
    if (process.env.NEXT_PUBLIC_VERCEL_ENV !== 'production') {
      headers.push({
        headers: [
          {
            key: 'X-Robots-Tag',
            value: 'noindex',
          },
        ],
        source: '/:path*',
      });
    }
    return headers;
  },
 };
 ```
